File: Python/Decorator.py
import abc;

# Interfaccia per l'oggetto base
class Component(abc.ABC):
	# Metodo comune, da implementare
	# Si può instanziare ???
	# def __new__(self):
	# 	### controllo se sottoclasse o meno?
	# 	pass;
	@abc.abstractmethod
	def operation(self):
		pass


# Classe astratta per decoratore
class Decorator(Component):
	
	# Serve rendere astratto il costruttore per non renderla instanziabile?
	# @abc.abstractmethod
	# def __init__(self, parent):
	#	pass
	
	def __init__(self, parent):
		self._parent = parent;
	
	# Metodo comune, ereditato
	# Inutile ridefinirlo astratto?
	@abc.abstractmethod
	def operation(self):
		self._parent.operation();


# Applicazione di classe base
class ConcreteComponent(Component):
	def operation(self):
		print("Sono la classe base");
	
	# Aggiungere qui altri metodi per la classe base


# Applicazione di decoratore
class ConcreteDecorator(Decorator):
	# Non dimenticare il costruttore!
	def ConcreteDecorator(self, parent):
		super().__init__(parent);
	
	# Reimplementazione non necessaria 
	def operation(self):
		print("Sono un decoratore. La classe che decoro diceva:");
		self._parent.operation();
	
	# Aggiungere qui altri metodi per il decoratore

File: Decorator.java
// Interfaccia per l'oggetto base
public interface Component() {
	// Metodo comune, da implementare
	void operation();
}

// Classe astratta per decoratore
public abstract class Decorator implements Component {
	protected Component parent;
	
	public Decorator(Component parent) {
		this.parent = parent;
	}
	
	// Metodo comune, ereditato
	public void operation() {
		parent.operation();
	}
}

// Applicazione di classe base
public class ConcreteComponent implements Component {
	@Override
	public void operation() {
		System.out.println("Sono la classe base");
	}
	
	// Aggiungere qui altri metodi per la classe base
}

// Applicazione di decoratore
public class ConcreteDecorator implements Decorator {
	// Non dimenticare il costruttore!
	public ConcreteDecorator(Component parent) {
		super(parent);
	}
	
	// Reimplementazione non necessaria 
	@Override
	public void operation() {
		System.out.println("Sono un decoratore. La classe che decoro diceva:");
		this.parent.operation();
	}
	
	// Aggiungere qui altri metodi per il decoratore
}